### DISCUSSION TITLE
Open Graph Crawler redirects

### DISCUSSION THREAD

**Discussion by coleUI** [vercel/vercel#6926] [Help]
I am trying to get Open Graph crawlers to get redirected to an endpoint on our server that returns the dynamic metadata for deep links. This works without the "has" key, but it redirects all traffic, even users. We just need this to redirect bots. The list of bot data is at https://github.com/monperrus/crawler-user-agents/blob/master/crawler-user-agents.json

`"rewrites": [
    {
      "source": "/proposals/:proposal_id/requests/:request_id",
      "has": [
        {
          "type": "header",
          "key": "Referer",
          "value": "https://api.slack.com/robots"
        },
        {
          "type": "header",
          "key": "Referer",
          "value": "https://dev.twitter.com/cards/getting-started"
        },
        {
          "type": "header",
          "key": "user-agent",
          "value": "Slackbot-LinkExpanding (+https://api.slack.com/robots)"
        },
        {
          "type": "header",
          "key": "user-agent",
          "value": "Slackbot-LinkExpanding 1.0 (+https://api.slack.com/robots)"
        },
        {
          "type": "header",
          "key": "User-Agent",
          "value": "Slackbot"
        },
        {
          "type": "header",
          "key": "user-agent",
          "value": "Slackbot"
        },
        {
          "type": "header",
          "key": "User-Agent",
          "value": "Twitterbot"
        },
        {
          "type": "header",
          "key": "user-agent",
          "value": "Twitterbot"
        }
      ],
      "destination": "https://server-dev.creaticles.com/creaticles-deep-link/requests/:request_id/proposals/:proposal_id"
    }
  ]`

**Comment by leerob** [comment]
Have you looked at Edge Functions and Middleware? I think this should more easily solve this problem by looking at the User-Agent and also being able to disallow specific traffic or users

https://github.com/vercel/examples/tree/main/edge-functions

**Comment by coleUI** [comment]
I have, but from what I have read, it requires converting our React app to
Nextjs, which we are hoping to avoid due to the close deadline. Is there a
way to use Edge Functions without converting over? Also, shouldn't the
vercel.json config also handle the redirects since we are setting the "has"
key to check for User Agent headers?

**Reply by leerob** [reply]
I would recommend using Middleware if you need to do more advanced logic than what the configuration allows for. You can use Middleware with normal React (without Next.js) by creating a `_middleware.js` file in the root of your directory.

**Comment by coleUI** [comment]
Can you please send any references on this? I had found this solution, but
I must have misinterpreted the docs.

Thanks

**Comment by ibastawisi** [comment]
This worked for me
```JSON
"redirects": [
    {
      "source": "/view/:id",
      "destination": "https://api.example.com/view/:id",
      "missing": [
        {
          "type": "header",
          "key": "Sec-Fetch-User"
        }
      ]
    }
  ]
  ```
The `Sec-Fetch-User` [fetch metadata request header](https://developer.mozilla.org/en-US/docs/Glossary/Fetch_metadata_request_header) is only sent for requests initiated by user activation
So if it is missing it must be a bot
